proposals/interface-types/CanonicalABI.md

```diff
@@ -0,0 +1,1200 @@
+  which lives for the duration of a function call (and can only show up as
+  paramters to functions). This enables efficient binding of APIs like `read`
+  and `write` where the canonical ABI is simply passing a pointer/length.
+
+With this grammar of interface types, we can now define what it means to lift
+and lower to all these types.
+
+## Core wasm signatures
+
+First let's take a look at the canonical core wasm signature for any particular
+interface types signature. Note that the canonical core wasm signatures for
+interface types only use the wasm MVP types `i32`, `i64`, `f32`, and `f64` (not
+the post-MVP types `externref`, `funcref`, and `v128`).
+
+```
+# Returns the core wasm signature for a function which takes `params` as
+# arguments (interface types) and returns `results`. The `direction` here is
+# either "import" or "export".
+#
+# Note the usage of a return pointer which is intended to be a temporary
+# workaround for the lack of multi-value today. Languages like C and Rust don't
+# have a great way of idiomatically working with multivalue returns just yet,
+# but this is likely to change in the near future.
+def wasm_signature(direction, params, results):
+  wasm_params = params.flat_map(|t| flatten(direction, t))
+  wasm_results = results.flat_map(|t| flatten(direction, t))
+
+  # If there's more than one result work around the lack of multi-value
+  # returns in C/Rust today. For imports the caller allocates a return pointer
+  # and for exports the callee returns a location in memory where everything
+  # was stored.
+  if len(wasm_results) > 1:
+    if direction == "import"
+      wasm_params.insert(0, i32)
+      wasm_results = []
+    else:
+      wasm_results = [i32]
+
+  return wasm_params, wasm_results
+
+def flatten(direction, interface_type):
+  match interface_type
+    f32 => [f32]
+    f64 => [f64]
+    s8 | s16 | s32 | u8 | u16 | u32 => [i32]
+    s64 | u64 => [i64]
+    char => [i32]
+
+    # List are represented by their pointer/length
+    string |
+    list $ty => [i32, i32]
+
+    # flags are represented as a contiguous list of `i32` values
+    flags $fields => {
+      fields = []
+      for _ in 0..num_packed_i32s($fields):
+        fields.push(i32)
+      fields
+    }
+
+    # Records have their fields flattened individually
+    record $fields => $fields.flat_map(|t| flatten(direction, t))
+
+    # variants are represented with a discriminant plus whatever is the
+    # smallest which fits all possible cases.
+    variant $cases => [i32] + flatten_cases(direction, $cases)
+
+    # Handles are represented as a single integer unconditionally. The
+    # un-forgeability is discussed later.
+    handle => [i32]
+
+    # Buffers are represented as a pointer/length when passing to an import
+    # and as an integer handle when receiving as part of an export.
+    push-buffer | pull-buffer =>
+      if direction == "import":
+        [i32, i32, i32]
```

Comment:
```suggestion
        [i32, i32]
```

? Or am I misunderstanding pointer/length encoding?

Reply:
Aha, it’s explained down the line:

    # The first i32 is either 0 meaning that this is a new local buffer within
    # `src`, or it's a nonzero value meaning that it's a handle to a buffer
    # defined in some other module.

might make sense to forward declare that comment here.